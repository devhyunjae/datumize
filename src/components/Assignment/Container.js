import styled from 'styled-components'

export default styled.div`
  margin-top: 15px;
  margin-bottom: 15px;
  width: 100%;
`
