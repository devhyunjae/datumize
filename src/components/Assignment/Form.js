import React, { PureComponent } from 'react'
import UserDropDown from './UsersDropdown'
import ProjectsDropDown from './ProjectsDropdown'
import RolesDropDown from './RolesDropdown'

class Form extends PureComponent {
  state = {
    selectedUser: null,
    selectedProject: null,
    selectedRole: null,
  }

  setSelectedUser = (selectedUser) => {
    this.setState({
      selectedUser,
    })
  }

  setSelectedProject = (selectedProject) => {
    this.setState({
      selectedProject,
    })
  }

  setSelectedRole = (selectedRole) => {
    this.setState({
      selectedRole,
    })
  }

  render() {
    return (
      <div>
        <UserDropDown setSelectedUser={this.setSelectedUser} />
        <ProjectsDropDown setSelectedProject={this.setSelectedProject} />
        <RolesDropDown setSelectedRole={this.setSelectedRole} />
        <button type="button" onClick={() => {alert('clicked')}}>Save</button>
      </div>
    )
  }
}

export default Form
