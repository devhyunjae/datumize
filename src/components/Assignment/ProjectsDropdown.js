import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import * as actions from '../../actions/projects'

class ProjectsDropdown extends PureComponent {
  static propTypes = {
    fetchProjects: PropTypes.func.isRequired,
    projects: PropTypes.array.isRequired,
    setSelectedProject: PropTypes.func,
  }

  static defaultProps = {
    setSelectedProject: () => {},
  }

  componentDidMount() {
    const { fetchProjects } = this.props
    fetchProjects()
  }

  componentDidUpdate() {
    const { projects, setSelectedProject } = this.props
    if (projects.length > 0) {
      const selectedProject = projects[0]
      setSelectedProject(selectedProject)
    }
  }

  changeEvent = (event) => {
    const { projects, setSelectedProject } = this.props
    const selectedProject = projects.find(project => project.id === parseInt(event.target.value, 0))
    setSelectedProject(selectedProject)
  }

  renderOptions() {
    const { projects } = this.props

    return projects.map(project => (
      <option value={project.id} key={project.id}>
        {project.name}
      </option>
    ))
  }

  render() {
    return (
      <div>
        <div>2. Choose a project</div>
        <div>
          <select onChange={this.changeEvent}>
            { this.renderOptions() }
          </select>
        </div>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  projects: state.projects.projects,
})

export { ProjectsDropdown }

export default connect(
  mapStateToProps,
  actions,
)(ProjectsDropdown)

