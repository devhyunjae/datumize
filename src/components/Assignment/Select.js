import React from 'react'
import PropTypes from 'prop-types'

const Select = ({ children }) => (
  <div className="pt-select pt-fill">
    { children }
  </div>
)

Select.propTypes = {
  children: PropTypes.element.isRequired,
}

export default Select
