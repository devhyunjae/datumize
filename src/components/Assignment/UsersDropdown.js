import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import * as actions from '../../actions/users'

class UsersDropdown extends PureComponent {
  static propTypes = {
    fetchUsers: PropTypes.func.isRequired,
    users: PropTypes.array.isRequired,
    setSelectedUser: PropTypes.func,
  }

  static defaultProps = {
    setSelectedUser: () => {},
  }

  componentDidMount() {
    const { fetchUsers } = this.props
    fetchUsers()
  }

  componentDidUpdate() {
    const { users, setSelectedUser } = this.props
    if (users.length > 0) {
      const selectedUser = users[0]
      setSelectedUser(selectedUser)
    }
  }

  changeEvent = (event) => {
    const { users, setSelectedUser } = this.props
    const selectedUser = users.find(user => user.id === parseInt(event.target.value, 0))
    setSelectedUser(selectedUser)
  }

  renderOptions() {
    const { users } = this.props

    return users.map(user => (
      <option value={user.id} key={user.id} name={user.name}>
        {user.name}
      </option>
    ))
  }

  render() {
    return (
      <div>
        <div>1. Choose a user</div>
        <div>
          <select onChange={this.changeEvent}>
            { this.renderOptions() }
          </select>
        </div>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  users: state.users.users,
})

export { UsersDropdown }

export default connect(
  mapStateToProps,
  actions,
)(UsersDropdown)

