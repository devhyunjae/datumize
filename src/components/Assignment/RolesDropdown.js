import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import * as actions from '../../actions/roles'
import Container from './Container'
import Select from './Select'

class RolesDropdown extends PureComponent {
  static propTypes = {
    fetchRoles: PropTypes.func.isRequired,
    roles: PropTypes.array.isRequired,
    setSelectedRole: PropTypes.func,
  }

  static defaultProps = {
    setSelectedRole: () => {},
  }

  componentDidMount() {
    const { fetchRoles } = this.props
    fetchRoles()
  }

  componentDidUpdate() {
    const { roles, setSelectedRole } = this.props
    if (roles.length > 0) {
      const selectedRole = roles[0]
      setSelectedRole(selectedRole)
    }
  }

  changeEvent = (event) => {
    const { roles, setSelectedRole } = this.props
    const selectedRole = roles.find(role => role.id === parseInt(event.target.value, 0))
    setSelectedRole(selectedRole)
  }

  renderOptions() {
    const { roles } = this.props

    return roles.map(role => (
      <option value={role.id} key={role.id}>
        {role.name}
      </option>
    ))
  }

  render() {
    return (
      <Container>
        <h5>3. Choose a role</h5>
        <Select>
          <select onChange={this.changeEvent}>
            { this.renderOptions() }
          </select>
        </Select>
      </Container>
    )
  }
}

const mapStateToProps = state => ({
  roles: state.roles.roles,
})

export { RolesDropdown }

export default connect(
  mapStateToProps,
  actions,
)(RolesDropdown)

