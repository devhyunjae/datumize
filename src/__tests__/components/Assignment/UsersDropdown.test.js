import React from 'react'
import renderer from 'react-test-renderer'
import { UsersDropdown } from '../../../components/Assignment/UsersDropdown'

describe('<UsersDropdown>', () => {
  it('renders correctly', () => {
    const fetchedUsers = [
      { id: 1, name: 'test01' },
      { id: 2, name: 'test02' },
    ]
    const component = renderer.create(
      <UsersDropdown
        fetchUsers={() => {}}
        users={fetchedUsers}
      />,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
