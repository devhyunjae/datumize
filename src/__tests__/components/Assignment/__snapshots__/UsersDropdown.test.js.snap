// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UsersDropdown> renders correctly 1`] = `
<div>
  <div>
    1. Choose a user
  </div>
  <div>
    <select
      onChange={[Function]}
    >
      <option
        name="test01"
        value={1}
      >
        test01
      </option>
      <option
        name="test02"
        value={2}
      >
        test02
      </option>
    </select>
  </div>
</div>
`;
